ast/impl/pretty_printer.cpp

```diff
@@ -0,0 +1,320 @@
+        node->parameters[i]->accept(*this);
+        if (i < node->parameters.size() - 1) {
+            fmt::print(", ");
+        }
+    }
+
+    fmt::print(")");
+
+    if (node->returnType != nullptr) {
+        fmt::print(" : ");
+
+        node->returnType->accept(*this);
+    }
+
+    m_oneLine--;
+
+    fmt::print(" is");
+
+    node->body->accept(*this);
+}
+
+void PrettyPrinter::visit(Type*) {
+    fmt::print("<stray type>\n", "", m_depth);
+}
+
+void PrettyPrinter::visit(AliasedType* node) {
+    fmt::print("{}", node->name);
+}
+
+void PrettyPrinter::visit(PrimitiveType*) {
+    fmt::print("<unknown primitive type>");
+}
+
+void PrettyPrinter::visit(IntegerType*) {
+    fmt::print("integer");
+}
+
+void PrettyPrinter::visit(RealType*) {
+    fmt::print("real");
+}
+
+void PrettyPrinter::visit(BooleanType*) {
+    fmt::print("boolean", "", m_depth);
+}
+
+void PrettyPrinter::visit(ArrayType* node) {
+    fmt::print("array");
+
+    if (node->length != nullptr) {
+        fmt::print(" [");
+        node->length->accept(*this);
+        fmt::print("]");
+    }
+
+    if (node->elementType != nullptr) {
+        fmt::print(" ");
+        node->elementType->accept(*this);
+    }
+}
+
+void PrettyPrinter::visit(RecordType* node) {
+    fmt::print("record");
+
+    if (!m_oneLine) {
+        m_depth++;
+    } else {
+        fmt::print(" ");
+    }
+
+    for (auto field : node->fields) {
+        if (!m_oneLine) {
+            newline();
+        }
+
+        field->accept(*this);
+
+        if (m_oneLine) {
+            fmt::print("; ");
+        }
+    }
+
+    if (!m_oneLine) {
+        m_depth--;
+        newline();
+        fmt::print("end");
+    } else {
+        fmt::print("end;");
+    }
+}
+
+void PrettyPrinter::visit(VariableDecl* node) {
+    fmt::print("var {}", node->name);
+
+    m_oneLine++;
+
+    if (node->type != nullptr) {
+        fmt::print(" : ");
+        node->type->accept(*this);
+    }
+
+    m_oneLine--;
+
+    if (node->initialValue != nullptr) {
+        fmt::print(" is ");
+        node->initialValue->accept(*this);
+    }
+}
+
+void PrettyPrinter::visit(TypeDecl* node) {
+    fmt::print("type {} is ", node->name);
+    
+    if (node->type != nullptr) {
+        node->type->accept(*this);
+    }
+}
+
+void PrettyPrinter::visit(Body* node) {
+    m_depth++;
+
+    for (auto type : node->types) {
+        newline();
+        type->accept(*this);
+        newline();
+    }
+
+    for (auto variableDecl : node->variables) {
+        newline();
+        variableDecl->accept(*this);
+    }
+
+    if (!node->variables.empty() && !node->statements.empty()) {
+        newline();
+    }
+
+    for (auto statement : node->statements) {
+        newline();
+        statement->accept(*this);
+    }
+
+    m_depth--;
+    newline();
+
+    // Yes, we'll have
+    // if x = 5 then
+    //     ...
+    // end else
+    //     ...
+    // end
+    // doesn't matter that much... I think
+    fmt::print("end");
```

Comment:
But the `end` keyword is generally not part of the body, it just highlights that it ends here. Better move printing `end` to the routine/if/for/while